marlowe-symbolic/src/Language/Marlowe/Analysis/FSSemantics.hs

```diff
@@ -1,1126 +1,415 @@
+               if (otherAccId == accId) && (otherParty == party)
+                  && (otherToken == token)
+               then (otherConcVal .== concVal) .|| previousMatch otherSymInput
+               else previousMatch otherSymInput
+             _ -> previousMatch otherSymInput
+     ((newCond, newSState), newTrace)
+               <- applyInputConditions (Just symInput) timeout sState pos cont
+     contTrace <- isValidAndFailsWhen rest timeout timCont newPreviousMatch sState (pos + 1)
+     return (ite (newCond .&& newTrace)
+                 newTrace
+                 ((newCond .&& (concVal .<= 0) -- Non-positive deposit warning
+                           .&& convertToSymbolicTrace (traces newSState)
+                                                      (paramTrace sState))
+                  .|| contTrace))
+isValidAndFailsWhen (Case (Choice choId bnds) cont:rest)
+                    timeout timCont previousMatch sState pos =
+  do concVal <- generateValueInBounds bnds
+     let symInput = SymChoice choId concVal
+     let clashResult = previousMatch symInput
+     let newPreviousMatch otherSymInput =
+           case otherSymInput of
+             SymChoice otherChoId otherConcVal ->
+               if otherChoId == choId
+               then (otherConcVal .== concVal) .|| previousMatch otherSymInput
+               else previousMatch otherSymInput
+             _ -> previousMatch otherSymInput
+     contTrace <- isValidAndFailsWhen rest timeout timCont newPreviousMatch sState (pos + 1)
+     ((newCond, newSState), newTrace)
+               <- applyInputConditions (Just symInput) timeout sState pos cont
+     return (ite (newCond .&& newTrace) newTrace contTrace)
+isValidAndFailsWhen (Case (Notify obs) cont:rest)
+                    timeout timCont previousMatch sState pos =
+  do let obsRes = symEvalObs obs sState
+     let symInput = SymNotify
+     let clashResult = previousMatch symInput
+     let newPreviousMatch otherSymInput =
+           case otherSymInput of
+             SymNotify -> sNot obsRes
+             _         -> previousMatch otherSymInput
+     (newCond, newSState) <- addTransaction (Just symInput) timeout sState pos
+     newTrace <- isValidAndFailsAux cont newSState
+     contTrace <- isValidAndFailsWhen rest timeout timCont newPreviousMatch sState (pos + 1)
+     return (ite (newCond .&& obsRes .&& newTrace) newTrace contTrace)
+
+-- SBV handling and result extraction
+
+countWhens :: Contract -> Integer
+countWhens Close               = 0
+countWhens (Pay uv uw ux uy c) = countWhens c
+countWhens (If uz c c2)        = max (countWhens c) (countWhens c2)
+countWhens (When cl t c)       = 1 + max (countWhensCaseList cl) (countWhens c)
+countWhens (Let va vb c)       = countWhens c
+
+countWhensCaseList :: [Case Contract] -> Integer
+countWhensCaseList (Case uu c : tail) = max (countWhens c) (countWhensCaseList tail)
+countWhensCaseList []                 = 0
```

Comment:
again in a `where`

Reply:
I would leave it in this case, again because it matches Isabelle semantics and having sub-functions there just makes it very hard to deal with.